pkg/parser/parser.go

```diff
@@ -587,14 +587,75 @@ func (p *Parser) parseBreakStatement(scope *scope) Node {
 	return breakStmt
 }
 
-// TODO: implemented.
 func (p *Parser) parseForStatement(scope *scope) Node {
-	scope = newScope(scope, nil)
+	forNode := &For{Token: p.cur}
+	scope = newScope(scope, forNode)
+	p.advance() // advance past FOR token
+
+	if p.cur.TokenType() != lexer.IDENT {
+		p.appendError("expected variable, found " + p.cur.FormatDetails())
+		p.advancePastNL()
+		return nil
+	}
+	forNode.LoopVar = &Var{Token: p.cur, Name: p.cur.Literal}
+	scope.set(forNode.LoopVar.Name, forNode.LoopVar)
+	p.advance() // advance past loopVarName
+
+	p.assertToken(lexer.DECLARE)
+	p.advance() // advance past :=
+	if !p.assertToken(lexer.RANGE) {
+		p.advancePastNL()
+		return nil
+	}
+	p.advance() // advance past range
+
+	n := p.parseExpr(scope, LOWEST)
+	if n == nil {
+		return nil
+	}
+	t := n.Type()
+	switch t.Name {
+	case STRING, MAP:
+		forNode.LoopVar.T = STRING_TYPE
+		forNode.Range = n
+		p.assertEOL()
+	case ARRAY:
+		forNode.LoopVar.T = t.Sub
+		forNode.Range = n
+		p.assertEOL()
+	case NUM:
+		forNode.LoopVar.T = NUM_TYPE
+		forNode.Range = p.parseStepRange(scope, n)
+	default:
+		p.appendError("expected num, string, array or map after range, found " + t.Format())
```

Comment:
On this branch, `forNode.LoopVar.T` is nil (the loopVar does not have a type). However, you go on to parse the block, and the code in that block can reference the var. That reference ends up failing because `forNode.LoopVar` (as set in the scope) is not a proper var, and you get a nil pointer dereference when trying to determine if the type is compatible when used.

Perhaps you could set the type to `any` for the purposes of continuing to parse the code? You may generate unrelated errors due to the type, but you won't crash.

Reply:
ok, I've set it to NONE_TYPE on generation. hope that will do.

Reply:
yeah that would be fine I expect. As long as it doesn't crash - it still doesn't compile - that error is correctly detected.